python/cuml/neighbors/kernel_density.py

```diff
@@ -0,0 +1,406 @@
+        #  the kernel density is sum(K(x,h)).
+        # In fact what they implment is (1/n)*sum(K(x,h))
+        # Here we divide by n in normal probability space
+        # Which becomes -log(n) in log probability space
+        sum_weights = (
+            cp.sum(self.sample_weight_)
+            if self.sample_weight_ is not None
+            else distances.shape[1]
+        )
+        log_probabilities -= np.log(sum_weights)
+
+        # norm
+        log_probabilities = norm_log_probabilities(
+            log_probabilities, self.kernel, h, X.shape[1]
+        )
+
+        return log_probabilities
+
+    def score(self, X, y=None):
+        """Compute the total log-likelihood under the model.
+
+        Parameters
+        ----------
+
+        X : array-like of shape (n_samples, n_features)
+            List of n_features-dimensional data points.  Each row
+            corresponds to a single data point.
+        y : None
+            Ignored.
+
+        Returns
+        -------
+
+        logprob : float
+            Total log-likelihood of the data in X. This is normalized to be a
+            probability density, so the value will be low for high-dimensional
+            data.
+        """
+        return cp.sum(self.score_samples(X))
+
+    def sample(self, n_samples=1, random_state=None):
+        """
+        Generate random samples from the model.
+        Currently, this is implemented only for gaussian and tophat kernels,
+        and the Euclidean metric.
+
+        Parameters
+        ----------
+        n_samples : int, default=1
+            Number of samples to generate.
+        random_state : int, cupy RandomState instance or None, default=None
+
+        Returns
+        -------
+        X : cupy array of shape (n_samples, n_features)
+            List of samples.
+        """
+        if not hasattr(self, "X_"):
+            raise NotFittedError()
+
+        if (self.kernel not in ["gaussian", "tophat"]
+                or self.metric != "euclidian"):
+            raise NotImplementedError()
+
+        if isinstance(random_state, cp.random.RandomState):
+            rng = random_state
+        else:
+            rng = cp.random.RandomState(random_state)
+
+        u = rng.uniform(0, 1, size=n_samples)
+        if self.sample_weight_ is None:
+            i = (u * self.X_.shape[0]).astype(np.int64)
+        else:
+            cumsum_weight = cp.cumsum(self.sample_weight_)
+            sum_weight = cumsum_weight[-1]
+            i = cp.searchsorted(cumsum_weight, u * sum_weight)
+        if self.kernel == "gaussian":
+            return cp.atleast_2d(rng.normal(self.X_[i], self.bandwidth))
+
+        elif self.kernel == "tophat":
+            # we first draw points from a d-dimensional normal distribution,
+            # then use an incomplete gamma function to map them to a uniform
+            # d-dimensional tophat distribution.
+            if not has_scipy():
```

Comment:
The `has_scipy()` function accepts an argument to raise an error if it's not available: https://github.com/rapidsai/cuml/blob/d2099b8ffc74163ca7bdca94dbbeb596081acf65/python/cuml/common/import_utils.py#L123